### PR TITLE
refactor: support batch transfer sudt with policy

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts [hunk 1]
```diff
@@ -4,75 +4,233 @@ import { Amount, Builder, Cell, RawTransaction, Transaction } from '@lay2/pw-cor
...
   readonly recipient: Address;
   readonly sudt: CkbTypeScript;
-  readonly amount: HexNumber;
+  amount: HexNumber;
+  policy: 'findAcp' | 'createCell' | 'findOrCreate';
 }
 
 export class TransferSudtPwBuilder extends AbstractPwSenderBuilder {
-  constructor(private options: TransferOptions, provider: CkitProvider, private sender: Address) {
+  constructor(private options: TransferOptions[], provider: CkitProvider, private sender: Address) {
```
Comment (homura) [inline packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]: to keep the interface as consistent as possible, maybe we should change options like this way

```ts
interface TransferOptions {
  recipients: TransferSudtRecipient[];
}

```

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts [hunk 2]
```diff
@@ -4,75 +4,233 @@ import { Amount, Builder, Cell, RawTransaction, Transaction } from '@lay2/pw-cor
...
+              amount: options.amount,
+              sudt: options.sudt,
+              policy: options.policy,
+            });
+      optionsGroupBySudt.set(options.sudt.args, newRecipientOptions);
+    });
+
+    const senderSudtInputCells = new Array<Cell>(0);
+    const recipientSudtInputCells = new Array<Cell>(0);
+    const senderSudtOutputCells = new Array<Cell>(0);
```
Comment (homura) [inline packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]: this is more common uses in ts

```ts
 const senderSudtInputCells: Cell[] = [];
```

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts [hunk 3]
```diff
@@ -4,75 +4,233 @@ import { Amount, Builder, Cell, RawTransaction, Transaction } from '@lay2/pw-cor
...
 
-    const senderOutput = senderInputs.reduce((sum, input) => {
-      sum.capacity = sum.capacity.add(input.capacity);
-      sum.setSUDTAmount(sum.getSUDTAmount().add(input.getSUDTAmount()));
-      return sum;
-    });
+    if (needSupplyCapacity) {
+      const outputsNeededCapacity = outputsContainedCapacity
+        // additional 61 ckb to ensure capacity is enough for change cell
+        .add(new Amount(String(byteLenOfCkbLiveCell())))
```
Comment (homura) [inline packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]:
```ts
byteLenOfCkbLiveCell(senderLockArgsLen);
```

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts [hunk 4]
```diff
@@ -4,75 +4,233 @@ import { Amount, Builder, Cell, RawTransaction, Transaction } from '@lay2/pw-cor
...
+            );
+
+            recipientSudtInputCells.push(recipientSudtInputCell);
+            recipientSudtOutputCells.push(recipientSudtOutputCell);
+            break;
+          }
+          case 'createCell': {
+            containCreateOption = true;
+            const recipientSudtOutputCell = new Cell(
+              new Amount('1').add(new Amount(String(byteLenOfSudt()))),
```
Comment (homura) [inline packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]:
```ts
new Amount('1').add(new Amount(String(byteLenOfSudt(recipientLockArgsLen))))
```

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts [hunk 5]
```diff
@@ -4,75 +4,233 @@ import { Amount, Builder, Cell, RawTransaction, Transaction } from '@lay2/pw-cor
...
+          }
+        }
+      }
+
+      const senderLiveSudtCells = await this.provider.collectUdtCells(
+        this.sender,
+        sudtScript,
+        totalTransferAmount.toHexString(),
+      );
+      const senderSudtInputCell: Cell[] = senderLiveSudtCells.map(
```
Comment (homura) [inline packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]:
```ts
const collectedSenderInputCells = senderLiveSudtCells.map(Pw.toPwCell);
```

- the `senderSudtInputCell` looks like a single cell, maybe we should change it
- Pw.toPwCell can convert collected cells to PwCell

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts (hunk 1 above)
Comment (ghost) [reply packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]: fixed

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts (hunk 2 above)
Comment (ghost) [reply packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]: fixed

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts (hunk 3 above)
Comment (ghost) [reply packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]: fixed

File: packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts (hunk 5 above)
Comment (ghost) [reply packages/ckit/src/tx-builders/pw/TransferSudtPwBuilder.ts]: fixed

**Pull request merged**